embedded/sql/stmt.go

```diff
@@ -845,6 +1092,36 @@ func (stmt *SelectStmt) Limit() uint64 {
 	return stmt.limit
 }
 
+func (stmt *SelectStmt) InferParameters(e *Engine) (map[string]SQLValueType, error) {
+	params := make(map[string]SQLValueType, 0)
+
+	err := stmt.inferParameters(e, e.implicitDB, params)
+	if err != nil {
+		return nil, err
+	}
+
+	return params, nil
+}
+
+func (stmt *SelectStmt) inferParameters(e *Engine, implicitDB *Database, params map[string]SQLValueType) error {
+	_, _, _, err := stmt.CompileUsing(e, implicitDB, nil)
+	if err != nil {
+		return err
+	}
+
+	snapshot, err := e.Snapshot()
```

Comment:
Is it guaranteed that this snapshot will be the same when executing the query? What if the schema changes in the meantime?

Reply:
good point. Currently there is no catalog changes that could affect inferred types but even in such case, the query resolution may raise errors due to types mismatch. Annotated this topic so to be accordingly analyzed.